airflow/providers/google/cloud/transfers/facebook_ads_to_gcs.py

```diff
@@ -100,15 +99,16 @@ def __init__(
         self.facebook_conn_id = facebook_conn_id
         self.api_version = api_version
         self.fields = fields
-        self.params = params
         self.gzip = gzip
         self.impersonation_chain = impersonation_chain
 
     def execute(self, context: dict):
         service = FacebookAdsReportingHook(
             facebook_conn_id=self.facebook_conn_id, api_version=self.api_version
         )
-        rows = service.bulk_facebook_report(params=self.params, fields=self.fields)
+        rows = service.bulk_facebook_report(
+            params={k: v.resolve() for k, v in self.params.items()}, fields=self.fields
+        )
```

Comment:
Yeah, so _this_ `params` is a name clash, and shouldn't be using the _same_ params.

This needs to not be the same as `dag.params`/typed/ etc.

I.e. this change needs to be  reverted.

Reply:
Yeah, these 2 operators should have named it differently, but wasn't. So if we revert this change the `params` won't get converted as `ParamsDict` and break things. However, if we keep this change, things still work fine as `BaseOperator.params` take place of `self.params`.

Reply:
This use of params was never a ParamsDict, and shouldn't be. This is only meant to be the params explicitly passed to this operator.

Reply:
Yup agree, but we can't put checks in 10 diff places for just these 2 operators. For example, dag.py's `validate_schedule_and_params` method would complain about `self.params` not being a ParamsDict.

Reply:
`validate_schedule_and_params` is on the DAG, not the operator/task

We can't change this use like this anyway -- it toally changes the behaviour of the operator

Reply:
Just tested, task params do contain the dag level params as well, so it would mix up simple params dict with a ParamsDict object and things would break. On the other hand, I'm still not clear why using base params instead of class-level params is not a good solution? Agree that the intended use would have been different, but as they messed up the naming, this solution seems to be a decent fix. About changing the behavior of operators, it won't, unless someone tries to do something fancy.

Reply:
Oh right, gotcha.

I think the main  thing here is that `dag.params` is so rarely used, that it wasn't the "intended" behvaiour here, so this is likely a bug that just no one has hit.

My main concern with this change is that because the providers are released separately this would create a versioning nightmare where the old provider can't work with new Airflow and vice-versa.

For example (all of these assuming a dag that does use `params):

- If you have the current provider and install  airflow 2.2, then `task.params` would end up with Param objects in them, not the "actual" values
- If you have Airflow 2.1 and install the next version of the provider that includes this change, you'd get an error trying to call `resolve()` on a `AttributeError: 'str' object has no attribute 'resolve'` (for example).

So is it possible to handle this entirely in core Airflow, not in the operators, so that by the time `operator.execute()` gets it's hand on `self.params` it is just the values, not the Param etc?

Reply:
We can easily add "additional_requirements' airflow `2.2.0+` for those two providers, to not complicate things. That will handle 2.2 with "hard" limit and it constraints will take care about the pre-2.2 default installation. We can also add a check in those new provider's code to raise warning/exception if they are used in 2.1 environment.

Reply:
We can do that, but I think in this particular case not having to deal with this at all in the Operator is a better solution.

Reply:
@ashb I would be more than happy to not touch these operators, but unfortunately, we've to. Couldn't see if there is any other workaround possible here.

Reply:
@msumit By resolving the `task.params` to a plain dict somewhere in this function https://github.com/apache/airflow/blob/2aa443f8d5f0fe9b4b37b9029e9bac5176c563ce/airflow/models/taskinstance.py#L1392-L1416

Reply:
BTW. The compatibility checker for providers: https://github.com/apache/airflow/pull/18020

Reply:
@potiuk :+1: Cool, that'll be useful for other reasons.

Reply:
@ashb I had exactly the same code before introducing `ParamsDict`. The whole reason to introduce `ParamsDict` was to make the dag or task params auto resolved using the dunder methods, so now going back to the old way defeats the whole purpose. Also, in that method, we've to put an `if` condition to check if it's a string or param object and then call `resolve` on that, which is a bit ugly and keeps holes open for some other unwanted issues/misuses in the future.

Reply:
@msumit my point is that by the time the operator gets executed it should just receive the actual fully resolved params - it shouldn't have to care about type validation or anything like that as it should already be handled before execution.

Otherwise every single operator that wants to access params at execute time will need to know about ParamsDict and to call resolve. And this feels like an anti pattern too me

Reply:
@ashb I think this issue arises in these 2 operators cause they were (mis)using a variable. The intended use of params is to be used with a templated string and that is not broken here. I don't think that there is any other operator which would or should be using params in this way. It's like tomorrow some operator uses dag's `start_date` in some weird way and Airflow has to honor that in the future?

Reply:
Yes, they were "misusing" the params, but what about example?

```python
@task
def my_op(params):
   assert isinstance(params['my_param'], str)
```

That should still be true (assuming `'my_param': Param(type='string', default='a')`.

My point is that I cant' think of any reason for an Operator's execute function to ever get anything than the resolved values.